File: Memorize/CardViewerViewController.m
//
//  CardViewerViewController.m
//  Memorize
//
//  Created by Nick Shelley on 4/20/14.
//  Copyright (c) 2014 Mine. All rights reserved.
//

#import "CardViewerViewController.h"
#import "Card.h"
#import "TextEditorViewController.h"

@interface CardViewerViewController ()
@property (weak, nonatomic) IBOutlet UITextView *questionTextView;
@property (weak, nonatomic) IBOutlet UITextView *answerTextView;

@end

@implementation CardViewerViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    UITapGestureRecognizer *questionTap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(questionTapped)];
    [questionTap setNumberOfTapsRequired:1];
    [self.questionTextView addGestureRecognizer:questionTap];
    
    UITapGestureRecognizer *answerTap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(answerTapped)];
    [answerTap setNumberOfTapsRequired:1];
    [self.answerTextView addGestureRecognizer:answerTap];
    
    self.questionTextView.text = self.existingCard.question ?: @"Touch to edit question.";
    self.answerTextView.text = self.existingCard.answer ?: @"Touch to edit answer.";
}

- (void)questionTapped {
    [self performSegueWithIdentifier:@"TextEditorSegue" sender:@"question"];
}

- (void)answerTapped {
    [self performSegueWithIdentifier:@"TextEditorSegue" sender:@"answer"];
}

- (void)editQuestion:(NSString *)text {
    if (!self.existingCard) {
        self.existingCard = [[Card alloc] init];
    }
    
    [self.existingCard updateQuestion:text];
    self.questionTextView.text = text;
}

- (void)editAnswer:(NSString *)text {
    if (!self.existingCard) {
        self.existingCard = [[Card alloc] init];
    }
    
    [self.existingCard updateAnswer:text];
    self.answerTextView.text = text;
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"TextEditorSegue"]) {
        TextEditorViewController *editor = segue.destinationViewController;
        if ([sender isEqualToString:@"question"]) {
            editor.existingText = self.existingCard.question;
            editor.saveBlock = ^void(NSString *text) {
                [self editQuestion:text];
            };
        } else {
            editor.existingText = self.existingCard.answer;
            editor.saveBlock = ^void(NSString *text) {
                [self editAnswer:text];
            };
        }
    }
}

@end

File: Memorize/CardViewerViewController.swift
//
//  CardViewerViewController.swift
//  Memorize
//
//  Created by Heather Shelley on 2/25/17.
//  Copyright © 2017 Mine. All rights reserved.
//

import UIKit

class CardViewerViewController: UIViewController {
    var existingCard: Card?
    
    @IBOutlet weak var questionTextView: UITextView?
    @IBOutlet weak var answerTextView: UITextView?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let questionTap = UITapGestureRecognizer(target: self, action: #selector(questionTapped))
        questionTap.numberOfTapsRequired = 1
        questionTextView?.addGestureRecognizer(questionTap)
        let answerTap = UITapGestureRecognizer(target: self, action: #selector(answerTapped))
        answerTap.numberOfTapsRequired = 1
        answerTextView?.addGestureRecognizer(answerTap)
        questionTextView?.text = existingCard?.question ?? "Touch to edit question."
        answerTextView?.text = existingCard?.answer ?? "Touch to edit answer."
    }
    
    func questionTapped() {
        performSegue(withIdentifier: "TextEditorSegue", sender: "question")
    }
    
    func answerTapped() {
        performSegue(withIdentifier: "TextEditorSegue", sender: "answer")
    }
    
    func editQuestion(_ text: String) {
        if existingCard == nil {
            existingCard = Card()
        }
        existingCard?.updateQuestion(text)
        questionTextView?.text = text
    }
    
    func editAnswer(_ text: String) {
        if existingCard == nil {
            existingCard = Card()
        }
        existingCard?.updateAnswer(text)
        answerTextView?.text = text
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        guard segue.identifier == "TextEditorSegue", let type = sender as? String, let editor = segue.destination as? TextEditorViewController else { return }
        
        if (type == "question") {
            editor.existingText = self.existingCard?.question
            editor.didSave = { text in
                self.editQuestion(text)
            }
        } else {
            editor.existingText = self.existingCard?.answer
            editor.didSave = { text in
                self.editAnswer(text)
            }
        }
    }
}
